subsys/bluetooth/controller/ll_sw/ull_df.c

```diff
@@ -665,4 +648,154 @@ static struct lll_df_adv_cfg *df_adv_cfg_acquire(void)
 
 	return df_adv_cfg;
 }
+
+/*
+ * @brief Function sets content of cte_info field in periodic advertising chain.
+ *
+ * The function allocates new PDU (or chain of PDUs) for periodic advertising to
+ * fill it with information about CTE that is going to be transmitted with the PDU.
+ * If there is already allocated PDU, it will be updated to hold information about
+ * CTE.
+ *
+ * @param adv            Pointer to periodic advertising set.
+ * @param df_cfg         Pointer to direction finding configuration
+ * @param[out] ter_idx   Pointer used to return index of allocated or updated PDU.
+ *                       Index is required for scheduling the PDU for transmission in LLL.
+ * @param[out] first_pdu Pointer to return address of first PDU in a chain
+ *
+ * @return Zero in case of success, other value in case of failure.
+ */
+static uint8_t set_cte_info_in_per_adv_chain(struct ll_adv_set *adv, struct lll_df_adv_cfg *df_cfg,
+					     uint8_t *ter_idx, struct pdu_adv **first_pdu)
+{
+	struct pdu_adv *pdu_prev, *pdu;
+	struct lll_adv_sync *lll_sync;
+	struct pdu_cte_info cte_info;
+	uint8_t pdu_add_field_flags;
+	void *extra_data;
+	void *hdr_data;
+	uint8_t err;
+
+	lll_sync = adv->lll.sync;
+
+	cte_info.type = df_cfg->cte_type;
+	cte_info.time = df_cfg->cte_length;
+	hdr_data = (uint8_t *)&cte_info;
+
+	/* Note: ULL_ADV_PDU_HDR_FIELD_CTE_INFO is just information that extra_data
+	 * is required in case of this ull_adv_sync_pdu_alloc.
+	 * Other flags here do not change anything. It may be changed to use
+	 * true/false flag for extra_data allocation.
+	 */
+	err = ull_adv_sync_pdu_alloc(adv, ULL_ADV_PDU_EXTRA_DATA_ALLOC_ALWAYS, &pdu_prev, &pdu,
+				     NULL, &extra_data, ter_idx);
+	if (err != BT_HCI_ERR_SUCCESS) {
+		return err;
+	}
+
+	ull_adv_sync_extra_data_set_clear(NULL, extra_data, ULL_ADV_PDU_HDR_FIELD_CTE_INFO, 0,
+					  df_cfg);
+
+#if defined(CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX) && (CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1)
+	if (df_cfg->cte_count > 1) {
+		pdu_add_field_flags =
+			ULL_ADV_PDU_HDR_FIELD_CTE_INFO | ULL_ADV_PDU_HDR_FIELD_AUX_PTR;
+	} else
+#endif /* CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX && CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1 */
+	{
+		pdu_add_field_flags = ULL_ADV_PDU_HDR_FIELD_CTE_INFO;
+	}
+
+	err = ull_adv_sync_pdu_set_clear(lll_sync, pdu_prev, pdu, pdu_add_field_flags, 0,
+					 hdr_data);
+	if (err != BT_HCI_ERR_SUCCESS) {
+		return err;
+	}
+
+	*first_pdu = pdu;
+
+#if defined(CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX) && (CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX > 1)
```

Comment:
The `CONFIG_BT_CTLR_DF_PER_ADV_CTE_NUM_MAX` should be guarded by `CONFIG_BT_CTLR_ADV_PDU_LINK`. I'll fix that. It can't be removed because the function uses `lll_adv_pdu_linked_next_get` function that is compiled conditionally.

Reply:
I've extracted appropriated part of code and created separate function.

Reply:
> It can't be removed because the function uses lll_adv_pdu_linked_next_get function that is compiled conditionally.

As long as the function has a header, it doesn't matter if it is compiled or not, as that would be optimized/handled by the linker.

Reply:
Yes, in this case linker will complain about unresolved `lll_adv_pdu_linked_next_get` reference.

Reply:
Then you must also be guarding `lll_adv_pdu_linked_next_get`, right? 

It's fine to keep as is, so no need to spend more time on that.